server/cluster/diagnostic_manager.go

```diff
@@ -0,0 +1,247 @@
+	}
+	return &diagnosticManager{
+		cluster:   cluster,
+		recorders: recorders,
+	}
+}
+
+func (d *diagnosticManager) getDiagnosticResult(name string) (*DiagnosticResult, error) {
+	isSchedulerExisted, _ := d.cluster.IsSchedulerExisted(name)
+	isDisabled, _ := d.cluster.IsSchedulerDisabled(name)
+	if !isSchedulerExisted || isDisabled {
+		ts := uint64(time.Now().Unix())
+		res := &DiagnosticResult{Name: name, Timestamp: ts, Status: disabled}
+		return res, nil
+	}
+
+	recorder := d.getRecorder(name)
+	if recorder == nil {
+		return nil, errs.ErrSchedulerUndiagnosable.FastGenByArgs(name)
+	}
+	result := recorder.getLastResult()
+	if result == nil {
+		return nil, errs.ErrNoDiagnosticResult.FastGenByArgs(name)
+	}
+	return result, nil
+}
+
+func (d *diagnosticManager) getRecorder(name string) *diagnosticRecorder {
+	return d.recorders[name]
+}
+
+// diagnosticRecorder is used to manage diagnose mechanism
+type diagnosticRecorder struct {
+	schedulerName string
+	cluster       *RaftCluster
+	summaryFunc   plan.Summary
+	results       *resultCache
+}
+
+func newDiagnosticRecorder(name string, cluster *RaftCluster) *diagnosticRecorder {
+	summaryFunc, ok := SummaryFuncs[name]
+	if !ok {
+		log.Error("can't find summary function", zap.String("scheduler-name", name))
+		return nil
+	}
+	return &diagnosticRecorder{
+		cluster:       cluster,
+		schedulerName: name,
+		summaryFunc:   summaryFunc,
+		results:       newResultMemorizer(maxDiagnosticResultNum),
+	}
+}
+
+func (d *diagnosticRecorder) isAllowed() bool {
+	if d == nil {
+		return false
+	}
+	return d.cluster.opt.IsDiagnosticAllowed()
+}
+
+func (d *diagnosticRecorder) getLastResult() *DiagnosticResult {
+	if d.results.Len() == 0 {
+		return nil
+	}
+	return d.results.generateResult()
+}
+
+func (d *diagnosticRecorder) setResultFromStatus(status string) {
+	if d == nil {
+		return
+	}
+	result := &DiagnosticResult{Name: d.schedulerName, Timestamp: uint64(time.Now().Unix()), Status: status}
+	d.results.put(result)
+}
+
+func (d *diagnosticRecorder) setResultFromPlans(ops []*operator.Operator, plans []plan.Plan) {
+	if d == nil {
+		return
+	}
+	result := d.analyze(ops, plans, uint64(time.Now().Unix()))
+	d.results.put(result)
+}
+
+func (d *diagnosticRecorder) analyze(ops []*operator.Operator, plans []plan.Plan, ts uint64) *DiagnosticResult {
+	res := &DiagnosticResult{Name: schedulers.BalanceRegionName, Timestamp: ts, Status: normal}
+	name := d.schedulerName
+	// TODO: support more schedulers and checkers
+	switch name {
+	case schedulers.BalanceRegionName:
+		runningNum := d.cluster.GetOperatorController().OperatorCount(operator.OpRegion)
+		if runningNum != 0 || len(ops) != 0 {
+			res.Status = scheduling
+			return res
+		}
+		res.Status = pending
+		if d.summaryFunc != nil {
+			isAllNormal := false
+			res.StoreStatus, isAllNormal, _ = d.summaryFunc(plans)
+			if isAllNormal {
+				res.Status = normal
+			}
+		}
+		return res
+	default:
+	}
+	// TODO: save plan into result
```

Comment:
Is this TODO finished?

Reply:
We do not need to do this now. In the future, we will need to save the plan if we want to support the analysis of detailed information